### PR TITLE
Add timeout Middleware Http support

### PR DESCRIPTION
**Pull request by amitksingh1490**

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/http/Middleware.scala [hunk 1]
```diff
@@ -204,7 +215,12 @@ object Middleware extends HttpMiddlewares {
         }
     }
 
-  final class PartialMake[AOut](val unit: Unit) extends AnyVal {
+  final class PartialTimeout[BOut](val unit: Unit) extends AnyVal {
+    def apply[R, E, AIn, BIn, AOut](
+      duration: Duration,
+    ): Middleware[R with Clock, E, AIn, AOut, BIn, Option[BOut]] = Middleware.identity.timeout(duration)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Middleware.scala]: I don't think we need all the type params, except for `R`

File: zio-http/src/test/scala/zhttp/http/MiddlewareSpec.scala [hunk 2]
```diff
@@ -61,6 +61,28 @@ object MiddlewareSpec extends DefaultRunnableSpec with HExitAssertion {
         val app = Http.succeed(1) @@ mid
         assertM(app(()) <& TestClock.adjust(3 second))(equalTo("B"))
       } +
+      suite("timeOut") {
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/http/MiddlewareSpec.scala]:
```suggestion
      suite("timeout") {
```

File: zio-http/src/test/scala/zhttp/http/MiddlewareSpec.scala [hunk 3]
```diff
@@ -61,6 +61,28 @@ object MiddlewareSpec extends DefaultRunnableSpec with HExitAssertion {
...
+          testM("other middleware completes") {
+            val mid = Middleware.succeed('A').delay(1 second)
+            val app = Http.succeed(1) @@ mid @@ Middleware.timeout[Char](2 second)
+            assertM(app(()) <& TestClock.adjust(3 second))(equalTo(Some('A')))
+          } +
+          testM("timeout middleware timed out") {
+            val mid = Middleware.succeed('A').delay(2 second)
+            val app = Http.succeed(1) @@ mid @@ Middleware.timeout[Char](1 second)
+            assertM(app(()) <& TestClock.adjust(3 second))(isNone)
+          }
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/http/MiddlewareSpec.scala:74]: I think two test shouldn't be enough? 

We can simplify the tests 
```scala
val mid = Middleware.timeout(1 second)
val app = Http.succeed(1).delay(2 second)
...
```

File: zio-http/src/main/scala/zhttp/http/Middleware.scala [hunk 4]
```diff
@@ -103,6 +103,12 @@ sealed trait Middleware[-R, +E, +AIn, -BIn, -AOut, +BOut] { self =>
   ): Middleware[R1, E1, AIn1, BIn1, AOut1, BOut1] =
     Middleware.Race(self, other)
 
+  /**
+   * Times out the application with a None
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Middleware.scala]: improve scala doc.

File: zio-http/src/main/scala/zhttp/http/Middleware.scala [hunk 5]
```diff
@@ -176,6 +182,11 @@ object Middleware extends HttpMiddlewares {
    */
   def succeed[B](b: B): Middleware[Any, Nothing, Nothing, Any, Any, B] = fromHttp(Http.succeed(b))
 
+  /**
+   * Times out if execution is not completed within specified duration, with a None output
+   */
+  def timeout[B]: PartialTimeout[B] = new PartialTimeout[B](())
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Middleware.scala:188]: I think we can drop the type param `B` and it should still be able to infer correctly.

Comment (amitksingh1490) [reply zio-http/src/main/scala/zhttp/http/Middleware.scala:188]: Not Infering

Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/http/Middleware.scala:188]: Let's drop this operator in that case. Until we are clear with the API signature we don't need to add it.

File: zio-http/src/main/scala/zhttp/http/middleware/HttpMiddlewares.scala [hunk 6]
```diff
@@ -155,7 +155,7 @@ private[zhttp] trait HttpMiddlewares extends CorsMiddlewares with CsrfMiddleware
   /**
    * Times out the application with a 408 status code.
    */
-  def timeout(duration: Duration): HttpMiddleware[Clock, Nothing] =
+  def timeoutRequest(duration: Duration): HttpMiddleware[Clock, Nothing] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/middleware/HttpMiddlewares.scala:158]: Revert to `timeout` 

Apologies for the to and fro.

**Comment by amitksingh1490** [comment]
Closing this PR not needed at current point of time

**Pull request closed without merging**